File: 0367. Valid Perfect Square/valid_perfect_square.cpp
class Solution
{
public:
    bool isPerfectSquare(int num);
private:
    int sqrtByLinearSearch(int n);
    int sqrtByBinarySearch(int n);
    int sqrtByNewton(int n);
};

bool Solution::isPerfectSquare(int num)
{
    int root = sqrtByNewton(num);
    return num % root == 0 && num / root == root;
}

int Solution::sqrtByLinearSearch(int n)
{
    int i = 1;
    while (i <= n / i)
        i++;
    return i - 1;
}

int Solution::sqrtByBinarySearch(int n)
{
    if (n == 1)
        return n;
    
    int left = 2, right = n / 2, mid;
    while (left <= right) {
        mid = left + ((right - left) >> 1);
        if (mid < n / mid)
            left = mid + 1;
        else if (mid > n / mid)
            right = mid - 1;
        else
            return mid;
    }
    return right;
}

int Solution::sqrtByNewton(int n)
{
    int r = n / 2;
    while (r > n / r)
        r = (r + n / r) / 2;
    return r;
}

File: 0367. Valid Perfect Square/valid_perfect_square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        r = self.sqrtByNewton(num)
        return num / r == r

    def sqrtByLinearSearch(self, n: int) -> int:
        i = 1
        while i <= n / i:
            i += 1
        return i - 1
        
    def sqrtByBinarySearch(self, n: int) -> int:
        left, right = 1, n // 2 + 1
        while left <= right:
            mid = left + (right - left) // 2
            if mid < n / mid:
                left = mid + 1
            elif mid > n / mid:
                right = mid - 1
            else:
                return mid
        return right

    def sqrtByNewton(self, n: int) -> int:
        r = n // 2 + 1
        while r > n / r:
            r = (r + n // r) // 2
        return r
